### ISSUE TITLE
Add KCP tunneling option

### ISSUE THREAD

**Issue by ghost** [Jigsaw-Code/outline-server#177]
I suggest adding an option to use KCP to encapsulate Shadowsocks traffic. This will help mobile and China users.

**Comment by sailosha** [comment]
KCPtun? It is cost more internet data usage to speed up transfer speed and lower the delay, well it is not necessary. I think we need tell new bird to use BBR to speed up first. Cuz in Asia cost of internet data transfer is not cheap. ( kcptun can boast up 5 times data usage, it gonna over cost)

I support we should give an opition to use it.

**Comment by studentmain** [comment]
China Mobile's router in some area can't handle kcp properly.

Yegor Ievlev <notifications@github.com> 于 2018年6月17日周日 02:24写道：

**Comment by YegorIevlev** [comment]
@studentmain In some countries UDP may be up to 100 Mbps and TCP is only 10 Mbps. In these cases KCP may be much faster.

**Comment by YegorIevlev** [comment]
On my current connection (100 Mbps advertised):
UDP: 81.9 Mbps
KCP: 51.4 Mbps
TCP: 8.28 Mbps

I got these values using iperf3 and a VPS by DigitalOcean located in Bangalore, India. I disabled compression and encryption and set all window sizes in kcptun to 1536.

**Comment by YegorIevlev** [comment]
@studentmain I don't mean we should enable it by default, just let the users decide.

**Comment by fortuna** [comment]
Please correct me if I'm wrong, but my understanding is the KCP has a very lax bandwidth control and achieves higher bandwidth by flooding the connection. While it will get higher speed, it can hurt other connections trying to use the network.

Besides being "unfair" to other clients, it also has a cost in terms of data and battery. You end up using the network card and sending a lot more data than you need to.

**Comment by fortuna** [comment]
I found this case where someone was seeing 60% of data waste with KCP: https://github.com/xtaci/kcptun/issues/441

When people measure KCP bandwidth, you have to be careful to measure the effective, useful, data transferred, and not the total data transferred, since there's a lot of redundancy and retransmission. The total data transferred is not the number we care about.

**Comment by arminmacx** [comment]
@fortuna As i see you really like to close the opened discussion without getting any result so fast (like mine and this about kcp). as @YegorIevlev said it could be decided features by user to enable it on server or not. As myself which already have my own shadowsocks+kcp for months i have better performance rather than using shadowsocks alone. So instead of closing it so fast lets see what other users thinking.

**Comment by fortuna** [comment]
@arminmacx I closed https://github.com/Jigsaw-Code/outline-server/issues/63 because it was a duplicate of this one, which remains open. It's better to discuss in one place, rather than spread the discussion in multiple places. I'm sorry for the confusion.

**Comment by fortuna** [comment]
More info (in English) about KCP:
https://github.com/skywind3000/kcp/blob/master/README.en.md

**Comment by arminmacx** [comment]
@fortuna i was oppend it before this one i guess because as i serched it i couldn't found any discussion about kcp that time

**Comment by fortuna** [comment]
We've seen huge boosts in performance in poor networks with BBR and this is corroborated by many reports online. So saying BBR sucks is simply wrong.

One other important thing to consider is that KCP is essentially a different protocol. You need extra software in the server and the client, and you lose compatibility. BBR is a modification of TCP, and anyone using Shadowsocks over TCP can still communicate without new software.

Now that we are done with the Windows full system VPN (phew!) we will take another look into adding BBR.

**Comment by sailosha** [comment]
Especially in Asia CN2 internet cost at least 0.2USD/GB, I set up on CN2 for the faster speed(boast up to 35Mb/s). BBR is super useful for this case, reduce delay and keep cost in the same level.

btw, @fortuna is outline already full system vpn?

**Comment by fortuna** [comment]
@sailosha , Outline has been a full VPN since launch in every platform it supported except for Windows. We launched the full VPN on Windows in September: https://www.reddit.com/r/outlinevpn/comments/9ivvkw/september_25th_whats_new_in_outline/

**Comment by sailosha** [comment]
@fortuna thanks so much for the reply!

**Comment by fortuna** [comment]
FYI, we have launched BBR for new DigitalOcean installations:
https://www.reddit.com/r/outlinevpn/comments/9rl82x/october_26th_bbr_support_on_new_digitalocean/

This can't be backported easily to older installations because the host image is an older version of Ubuntu, with a kernel that doesn't support BBR. If you're advanced, you can upgrade the host kernel and enable BBR: https://www.techrepublic.com/article/how-to-enable-tcp-bbr-to-improve-network-speed-on-linux/

**Comment by hktalent** [comment]
ump tunnel socks5?
@fortuna

**Comment by fortuna** [comment]
@hktalent I didn't understand your question. Outline Servers relay UDP traffic.

**Comment by Henry586** [comment]
I just setup Outline with chacha20-ietf-poly1305, but it was blocked by GFW after several days .
While Shadowrocket over Kcptun is still fine after then.

**Comment by jinixx** [comment]
Hi guys, doesn't seem to have any update ever since. Is kcptun still on the road map?

**Comment by fortuna** [comment]
KCP tunneling is not on the roadmap.

Without a clear story for how KCP helps we won't consider it further.

Take the [example of Shadowrocket over KCP](https://github.com/Jigsaw-Code/outline-server/issues/177#issuecomment-1043720914). Is the reason why it's not blocked is KCP or is it that UDP is not blocked?
If we have to use UDP, there are other options to consider.

KCP also has design issues. It floods the network, that's why it gets higher speed. But put multiple KCP users in the same network and you've degraded your network.

It seems the protocol allowlist in China is no longer enabled, so Outline Shadowsocks should work fine, if it gets re-enabled, you can [use TCP connection prefixes](https://www.reddit.com/r/outlinevpn/wiki/index/prefixing/) to bypass it.
